### PR TITLE
Add DNMD API to create a new metadata handle and implement IMetaDataDispenser::DefineScope

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Add a new API so we can create images with DNMD (and use them to create bespoke images for testing corner cases in metadata logic).

Also, implement `IMetaDataDispenser::DefineScope` with this API. Since we don't have an `IMetaDataEmit` implementation yet, it's not really useful as you get a read-only view to a newly created module, but once we add `IMetaDataEmit` support, it'll be very quick to enable the support for editing.

### CODE REVIEW

File: src/interfaces/dispenser.cpp [hunk 1]
```diff
@@ -24,11 +24,50 @@ namespace
...
+            }
+
+            mdhandle_ptr md_ptr { md_create_new_handle() };
+            if (md_ptr == nullptr)
+                return E_OUTOFMEMORY;
+            
+            // Initialize the MVID of the new image.
+            mdcursor_t moduleCursor;
+            if (!md_token_to_cursor(md_ptr.get(), TokenFromRid(1, mdtModule), &moduleCursor))
+                return E_UNEXPECTED;
```
Comment (jkoritzinsky) [inline src/interfaces/dispenser.cpp:45]:
```suggestion
                return E_FAIL;
```

File: src/dnmd/entry.c [hunk 2]
```diff
@@ -248,6 +248,89 @@ bool md_create_handle(void const* data, size_t data_len, mdhandle_t* handle)
...
+    uint32_t generation = 0;
+    if (1 != md_set_column_value_as_constant(module_cursor, mdtModule_Generation, 1, &generation))
+        return false;
+    
+    // Use the 0 index to specify the NULL guid as the guids for the image.
+    uint32_t guid_heap_offset = 0;
+    if (1 != set_column_value_as_heap_offset(module_cursor, mdtModule_Mvid, 1, &guid_heap_offset)
+        || 1 != set_column_value_as_heap_offset(module_cursor, mdtModule_EncBaseId, 1, &guid_heap_offset)
+        || 1 != set_column_value_as_heap_offset(module_cursor, mdtModule_EncId, 1, &guid_heap_offset))
+        return false;
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/entry.c:271]:
```suggestion
    if (1 != set_column_value_as_heap_offset(module_cursor, mdtModule_Mvid, 1, &guid_heap_offset)
        || 1 != set_column_value_as_heap_offset(module_cursor, mdtModule_EncBaseId, 1, &guid_heap_offset)
        || 1 != set_column_value_as_heap_offset(module_cursor, mdtModule_EncId, 1, &guid_heap_offset))
    {
        return false;
    }
```

File: src/dnmd/entry.c [hunk 3]
```diff
@@ -248,6 +248,89 @@ bool md_create_handle(void const* data, size_t data_len, mdhandle_t* handle)
...
+
+    const char* name = "";
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/entry.c:273]:
```suggestion
    char const* name = "";
```

File: src/dnmd/entry.c [hunk 4]
```diff
@@ -248,6 +248,89 @@ bool md_create_handle(void const* data, size_t data_len, mdhandle_t* handle)
...
+    if (1 != md_set_column_value_as_utf8(module_cursor, mdtModule_Name, 1, &name))
+        return false;
+    
+    md_commit_row_add(module_cursor);
+
+    mdcursor_t global_type_cursor;
+    if (!md_append_row(cxt, mdtid_TypeDef, &global_type_cursor))
+        return false;
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/entry.c:281]: Can you comment on the commit row add and the append row logic?

File: src/dnmd/entry.c [hunk 5]
```diff
@@ -248,6 +248,89 @@ bool md_create_handle(void const* data, size_t data_len, mdhandle_t* handle)
...
+
+    uint32_t flags = 0;
+    if (1 != md_set_column_value_as_constant(global_type_cursor, mdtTypeDef_Flags, 1, &flags))
+        return false;
+    
+    const char* global_type_name = "<Module>";
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/entry.c:287]:
```suggestion
    char const* global_type_name = "<Module>"; // Defined in ECMA-335 II.10.8
```

File: src/dnmd/entry.c [hunk 6]
```diff
@@ -248,6 +248,89 @@ bool md_create_handle(void const* data, size_t data_len, mdhandle_t* handle)
...
+    
+    if (!initialize_minimal_table_rows(pcxt))
+    {
+        free(pcxt);
+        return NULL;
+    }
+
+    return pcxt;
+}
+
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/entry.c:332]:
```suggestion
```

File: src/inc/dnmd.h [hunk 7]
```diff
@@ -38,6 +38,21 @@ typedef void* mdhandle_t;
...
+// and have an MVID of all zeros.
+mdhandle_t md_create_new_handle();
+
+#ifdef DNMD_PORTABLE_PDB
+// Create a new metadata handle for a new Portable PDB image.
+// Returns a handle for the new image, or NULL if the handle could not be created.
+// The image will always be in the v1.1 metadata format
+// and use the "PDB v1.0" version string.
+mdhandle_t md_create_new_pdb_handle();
+#endif
```
Comment (AaronRobinsonMSFT) [inline src/inc/dnmd.h:54]:
```suggestion
#endif // DNMD_PORTABLE_PDB
```

File: src/dnmd/entry.c [hunk 8]
```diff
@@ -248,6 +248,119 @@ bool md_create_handle(void const* data, size_t data_len, mdhandle_t* handle)
...
+    // Allocate and initialize a full context
+    // with the correctly-sized trailing memory.
+    mdcxt_t* pcxt = allocate_full_context(&cxt);
+    if (pcxt == NULL)
+        return NULL;
+
+    return pcxt;
+}
+#endif
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/entry.c:362]:
```suggestion
#endif // DNMD_PORTABLE_PDB
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**